smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/ResolveEndpointMiddlewareGenerator.kt

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package software.amazon.smithy.kotlin.codegen.rendering.endpoints
+
+import software.amazon.smithy.codegen.core.CodegenException
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.kotlin.codegen.KotlinSettings
+import software.amazon.smithy.kotlin.codegen.core.*
+import software.amazon.smithy.kotlin.codegen.model.*
+import software.amazon.smithy.kotlin.codegen.rendering.protocol.ProtocolGenerator
+import software.amazon.smithy.kotlin.codegen.utils.getOrNull
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.knowledge.OperationIndex
+import software.amazon.smithy.model.shapes.MemberShape
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.rulesengine.language.EndpointRuleSet
+import software.amazon.smithy.rulesengine.language.syntax.parameters.Parameter
+import software.amazon.smithy.rulesengine.language.syntax.parameters.ParameterType
+import software.amazon.smithy.rulesengine.traits.ClientContextParamsTrait
+import software.amazon.smithy.rulesengine.traits.ContextParamTrait
+import software.amazon.smithy.rulesengine.traits.StaticContextParamsTrait
+
+/**
+ * Generates resolver middleware for a specific endpoint provider signature.
+ * The binding of parameter values is generated separately in [EndpointParameterBindingGenerator], and that generated
+ * implementation is plugged in here.
+ */
+class ResolveEndpointMiddlewareGenerator(
+    private val ctx: ProtocolGenerator.GenerationContext,
+    private val writer: KotlinWriter,
+    private val renderPostResolution: () -> Unit = {},
+) {
+    companion object {
+        const val CLASS_NAME = "ResolveEndpointMiddleware"
+
+        fun getSymbol(settings: KotlinSettings): Symbol =
+            buildSymbol {
+                name = CLASS_NAME
+                namespace = "${settings.pkg.name}.endpoints.internal"
+            }
+    }
+
+    fun render() {
+        writer.openBlock("internal class #L<I, O>(", CLASS_NAME)
+        renderConstructorParams()
+        writer.closeAndOpenBlock(") : #T<I, O> {", RuntimeTypes.Http.Operation.InlineMiddleware)
+        renderClassMembers()
+        writer.write("")
+        renderInstall()
+        writer.closeBlock("}")
+    }
+
+    private fun renderConstructorParams() {
+        writer.write("private val endpointProvider: #T,", EndpointProviderGenerator.getSymbol(ctx.settings))
+        writer.write("private val buildParams: #T.Builder.(input: I) -> Unit,", EndpointParametersGenerator.getSymbol(ctx.settings))
```

Comment:
IDK why this irks me so much but it does. I don't have a better suggestion though so it's fine for now.


Part of my distaste for this is that it results in backing lambdas for each operation, e.g.

```
> javap DefaultS3Client\$deleteBucket\$2                  
Warning: File ./DefaultS3Client$deleteBucket$2.class does not contain class DefaultS3Client$deleteBucket$2
Compiled from "DefaultS3Client.kt"
final class aws.sdk.kotlin.services.s3.DefaultS3Client$deleteBucket$2 extends kotlin.jvm.internal.Lambda implements kotlin.jvm.functions.Function2<aws.sdk.kotlin.services.s3.endpoints.EndpointParameters$Builder, aws.sdk.kotlin.services.s3.model.DeleteBucketRequest, kotlin.Unit> {
  final aws.sdk.kotlin.services.s3.DefaultS3Client this$0;
  aws.sdk.kotlin.services.s3.DefaultS3Client$deleteBucket$2(aws.sdk.kotlin.services.s3.DefaultS3Client);
  public final void invoke(aws.sdk.kotlin.services.s3.endpoints.EndpointParameters$Builder, aws.sdk.kotlin.services.s3.model.DeleteBucketRequest);
  public java.lang.Object invoke(java.lang.Object, java.lang.Object);
}
```

The alternative isn't much better though as it would require us to generate a middleware per/operation anyway

Reply:
Yeah, I had the per-operation middleware before refactoring to this, which was much worse IMO.